src/components/molecules/Settings/Project/PublishSection/hooks.ts

```diff
@@ -1,63 +1,122 @@
-      if (!value) return;
-      navigator.clipboard.writeText(value);
-    },
-    [],
-  );
+export default (
+  defaultAlias?: string,
+  loading?: boolean,
+  publicationStatus?: Status,
+  validAlias?: boolean,
+  validatingAlias?: boolean,
+  onAliasValidate?: (alias: string) => void,
+  onPublish?: (alias: string | undefined, publicationStatus: Status) => void | Promise<void>,
+) => {
+  const { getAccessToken } = useAuth();
+  const url = window.REEARTH_CONFIG?.published?.split("{}");
+  const extensions = window.REEARTH_CONFIG?.extensions?.publication;
+
+  const [alias, setAlias] = useState(defaultAlias);
+  const [validation, setValidation] = useState<Validation>();
+  const [showDModal, setDModal] = useState(false);
+  const [accessToken, setAccessToken] = useState<string>();
 
-  const validate = useCallback(
+  const handleAliasValidate = useCallback(
     (a?: string) => {
       if (!a) {
-        changeValidation(undefined);
+        setValidation(undefined);
         return;
       }
       if (a.length < 5) {
-        changeValidation("too short");
+        setValidation("too short");
       } else if (!/^[A-Za-z0-9_-]*$/.test(a)) {
-        changeValidation("not match");
+        setValidation("not match");
       } else {
-        changeValidation(undefined);
+        setValidation(undefined);
         onAliasValidate?.(a);
       }
     },
     [onAliasValidate],
   );
 
-  const generateAlias = useCallback(() => {
+  const handleGenerateAlias = useCallback(() => {
     const str = generateRandomString(15);
-    changeAlias(str);
+    setAlias(str);
     return str;
   }, []);
 
-  const onAliasChange = useCallback(
+  const handleAliasChange = useCallback(
     (value?: string) => {
-      const a = value || generateAlias();
-      changeAlias(a);
-      validate(a);
+      const a = value || handleGenerateAlias();
+      setAlias(a);
+      handleAliasValidate(a);
     },
-    [validate], // eslint-disable-line react-hooks/exhaustive-deps
+    [handleAliasValidate], // eslint-disable-line react-hooks/exhaustive-deps
   );
 
+  const purl = useMemo(() => {
+    return (url?.[0] ?? "") + (defaultAlias?.replace("/", "") ?? "") + (url?.[1] ?? "");
+  }, [url, defaultAlias]);
+
+  const handleCopyToClipBoard = useCallback(
+    (value: string | undefined) => () => {
+      if (!value) return;
+      navigator.clipboard.writeText(value);
+    },
+    [],
+  );
+
+  const handlePublish = useCallback(async () => {
+    if (!publicationStatus) {
+      setDModal(false);
+      return;
+    }
+    await onPublish?.(alias, publicationStatus);
+    setDModal(false);
+  }, [alias, onPublish, publicationStatus]);
+
+  const publishDisabled = useMemo(
+    () =>
+      loading ||
+      publicationStatus === "unpublished" ||
+      (publicationStatus === "published" &&
+        (!alias || !!validation || validatingAlias || !validAlias)),
+    [alias, validation, validAlias, validatingAlias, publicationStatus, loading],
+  );
```

Comment:
```suggestion
  const publishDisabled = loading ||
      publicationStatus === "unpublished" ||
      (publicationStatus === "published" &&
        (!alias || !!validation || validatingAlias || !validAlias));
```

useMemo is not required because that always takes only constant calculation time and there is no reference to some objects.